Add a `UUID` specialization to `ObjectIdentifier`

rr-codes opened the pull request:
#### 3e6c9b9adb42cc075ecbe93cdf5a34c8c8762b72
<pre>
Add a `UUID` specialization to `ObjectIdentifier`
<a href="https://bugs.webkit.org/show_bug.cgi?id=274738">https://bugs.webkit.org/show_bug.cgi?id=274738</a>
<a href="https://rdar.apple.com/128772989">rdar://128772989</a>

Reviewed by Chris Dumez.

Allow `UUID`-backed `ObjectIdentifier`s to be created.

* Source/WTF/wtf/CrossThreadCopier.h:
* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/ObjectIdentifier.cpp:
(WTF::ObjectIdentifierMainThreadAccessTraits&lt;uint64_t&gt;::generateIdentifierInternal):
(WTF::ObjectIdentifierThreadSafeAccessTraits&lt;uint64_t&gt;::generateIdentifierInternal):
(WTF::operator&lt;&lt;):
(WTF::ObjectIdentifierMainThreadAccessTraits&lt;UUID&gt;::generateIdentifierInternal):
(WTF::ObjectIdentifierThreadSafeAccessTraits&lt;UUID&gt;::generateIdentifierInternal):
(WTF::ObjectIdentifierMainThreadAccessTraits::generateIdentifierInternal): Deleted.
(WTF::ObjectIdentifierThreadSafeAccessTraits::generateIdentifierInternal): Deleted.
* Source/WTF/wtf/ObjectIdentifier.h:
(WTF::ObjectIdentifierGenericBase&lt;uint64_t&gt;::toUInt64 const):
(WTF::ObjectIdentifierGenericBase&lt;uint64_t&gt;::toRawValue const):
(WTF::ObjectIdentifierGenericBase&lt;uint64_t&gt;::isValidIdentifier):
(WTF::ObjectIdentifierGenericBase&lt;uint64_t&gt;::ObjectIdentifierGenericBase):
(WTF::ObjectIdentifierGenericBase&lt;uint64_t&gt;::hashTableDeletedValue):
(WTF::ObjectIdentifierGenericBase&lt;UUID&gt;::isHashTableDeletedValue const):
(WTF::ObjectIdentifierGenericBase&lt;UUID&gt;::isValid const):
(WTF::ObjectIdentifierGenericBase&lt;UUID&gt;::toRawValue const):
(WTF::ObjectIdentifierGenericBase&lt;UUID&gt;::operator bool const):
(WTF::ObjectIdentifierGenericBase&lt;UUID&gt;::loggingString const):
(WTF::ObjectIdentifierGenericBase&lt;UUID&gt;::isValidIdentifier):
(WTF::ObjectIdentifierGenericBase&lt;UUID&gt;::ObjectIdentifierGenericBase):
(WTF::ObjectIdentifierGenericBase&lt;UUID&gt;::hashTableDeletedValue):
(WTF::ObjectIdentifierGeneric::ObjectIdentifierGeneric):
(WTF::add):
(WTF::ObjectIdentifierGenericBaseHash&lt;uint64_t&gt;::hash):
(WTF::ObjectIdentifierGenericBaseHash&lt;uint64_t&gt;::equal):
(WTF::ObjectIdentifierGenericBaseHash&lt;UUID&gt;::hash):
(WTF::ObjectIdentifierGenericBaseHash&lt;UUID&gt;::equal):
(WTF::operator==):
(WTF::operator&gt;):
(WTF::operator&gt;=):
(WTF::operator&lt;):
(WTF::operator&lt;=):
(WTF::ObjectIdentifierGenericBase::isHashTableDeletedValue const): Deleted.
(WTF::ObjectIdentifierGenericBase::isValid const): Deleted.
(WTF::ObjectIdentifierGenericBase::toUInt64 const): Deleted.
(WTF::ObjectIdentifierGenericBase::operator bool const): Deleted.
(WTF::ObjectIdentifierGenericBase::loggingString const): Deleted.
(WTF::ObjectIdentifierGenericBase::isValidIdentifier): Deleted.
(WTF::ObjectIdentifierGenericBase::ObjectIdentifierGenericBase): Deleted.
(WTF::ObjectIdentifierGenericBase::hashTableDeletedValue): Deleted.
(WTF::ObjectIdentifierGenericBaseHash::hash): Deleted.
(WTF::ObjectIdentifierGenericBaseHash::equal): Deleted.
(WTF::ObjectIdentifierGenericBaseStringTypeAdapter::length const): Deleted.
(WTF::ObjectIdentifierGenericBaseStringTypeAdapter::is8Bit const): Deleted.
(WTF::ObjectIdentifierGenericBaseStringTypeAdapter::writeTo const): Deleted.
(WTF::ObjectIdentifierGenericBaseStringTypeAdapter::ObjectIdentifierGenericBaseStringTypeAdapter): Deleted.
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::escapeIDForJSON):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendWithAsyncReply):
(IPC::Connection::sendWithPromisedReply):
(IPC::Connection::sendWithPromisedReplyOnDispatcher):
(IPC::Connection::send):
(IPC::Connection::sendSync):
(IPC::Connection::waitForAndDispatchImmediately):
* Source/WebKit/Platform/IPC/MessageSender.h:
* Source/WebKit/Platform/IPC/MessageSenderInlines.h:
(IPC::MessageSender::send):
(IPC::MessageSender::sendSync):
(IPC::MessageSender::sendWithAsyncReply):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::send):
(IPC::StreamClientConnection::sendWithAsyncReply):
(IPC::StreamClientConnection::sendSync):
(IPC::StreamClientConnection::waitForAndDispatchImmediately):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
(IPC::StreamServerConnection::send):
* Source/WebKit/Shared/AuxiliaryProcess.h:
(WebKit::AuxiliaryProcess::addMessageReceiver):
(WebKit::AuxiliaryProcess::removeMessageReceiver):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::sendWithAsyncReply):
(WebKit::AuxiliaryProcessProxy::send):
(WebKit::AuxiliaryProcessProxy::sendSync):
(WebKit::AuxiliaryProcessProxy::addMessageReceiver):
(WebKit::AuxiliaryProcessProxy::removeMessageReceiver):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::sendToAllProcesses):
* Source/WebKit/UIProcess/RemotePageProxy.h:
(WebKit::RemotePageProxy::sendWithAsyncReply):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::send):
(WebKit::RemoteRenderingBackendProxy::sendSync):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:

Canonical link: <a href="https://commits.webkit.org/279402@main">https://commits.webkit.org/279402@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac8f023c867b2c0e9f1448a967b177b7a8fcd0e5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53223 "2 style errors") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32563 "Hash ac8f023c for PR 29134 does not build (failure)") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5712 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56502 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3949 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39801 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3686 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43146 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2568 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55321 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30591 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45954 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24277 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27538 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3280 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2105 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46579 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49322 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3437 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58098 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52736 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28368 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3404 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50548 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29584 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46175 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49866 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11636 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30505 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65041 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29342 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12384 "Found 15 new JSC stress test failures: stress/proxy-get-prototype-of.js.bytecode-cache, stress/proxy-set-prototype-of.js.mini-mode, stress/sampling-profiler-display-name.js.dfg-eager, wasm.yaml/wasm/function-tests/br-table-as-return.js.wasm-eager, wasm.yaml/wasm/function-tests/brTableAsIf.js.wasm-eager, wasm.yaml/wasm/function-tests/brTableManyValues.js.wasm-eager, wasm.yaml/wasm/function-tests/float-sub.js.wasm-eager, wasm.yaml/wasm/function-tests/loop-sum.js.wasm-eager, wasm.yaml/wasm/function-tests/shl.js.wasm-eager, wasm.yaml/wasm/js-api/Module-compile.js.wasm-eager ... (failure)") | 
<!--EWS-Status-Bubble-End-->

megangardner commented on Source/WTF/wtf/ObjectIdentifier.h:
```diff
@@ -26,37 +26,66 @@
     bool isValid() const { return isValidIdentifier(m_identifier); }
 
-    uint64_t toUInt64() const { return m_identifier; }
+    RawValue toUInt64() const { return m_identifier; } // Use `toRawValue` instead.
```
I don't understand the above comment.

megangardner replied:
Also this seems wrong? Why have a RawValue return for a toUInt64 function?

rr-codes replied:
RawValue is just a type alias for `uint64_t` for this specialization, which promotes consistency between the specializations. The comment means that you should use `toRawValue` instead of `toUint64` since even though they both do the same thing, the former again helps consistency. Ideally, I would remove `toUint64`, but it's used in a lot of places, so this is the next best design.

cdumez reviewed (approved):
LGTM

rr-codes committed:
Add a `UUID` specialization to `ObjectIdentifier`
https://bugs.webkit.org/show_bug.cgi?id=274738
rdar://128772989

Reviewed by Chris Dumez.

Allow `UUID`-backed `ObjectIdentifier`s to be created.

* Source/WTF/wtf/CrossThreadCopier.h:
* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/ObjectIdentifier.cpp:
(WTF::ObjectIdentifierMainThreadAccessTraits<uint64_t>::generateIdentifierInternal):
(WTF::ObjectIdentifierThreadSafeAccessTraits<uint64_t>::generateIdentifierInternal):
(WTF::operator<<):
(WTF::ObjectIdentifierMainThreadAccessTraits<UUID>::generateIdentifierInternal):
(WTF::ObjectIdentifierThreadSafeAccessTraits<UUID>::generateIdentifierInternal):
(WTF::ObjectIdentifierMainThreadAccessTraits::generateIdentifierInternal): Deleted.
(WTF::ObjectIdentifierThreadSafeAccessTraits::generateIdentifierInternal): Deleted.
* Source/WTF/wtf/ObjectIdentifier.h:
(WTF::ObjectIdentifierGenericBase<uint64_t>::toUInt64 const):
(WTF::ObjectIdentifierGenericBase<uint64_t>::toRawValue const):
(WTF::ObjectIdentifierGenericBase<uint64_t>::isValidIdentifier):
(WTF::ObjectIdentifierGenericBase<uint64_t>::ObjectIdentifierGenericBase):
(WTF::ObjectIdentifierGenericBase<uint64_t>::hashTableDeletedValue):
(WTF::ObjectIdentifierGenericBase<UUID>::isHashTableDeletedValue const):
(WTF::ObjectIdentifierGenericBase<UUID>::isValid const):
(WTF::ObjectIdentifierGenericBase<UUID>::toRawValue const):
(WTF::ObjectIdentifierGenericBase<UUID>::operator bool const):
(WTF::ObjectIdentifierGenericBase<UUID>::loggingString const):
(WTF::ObjectIdentifierGenericBase<UUID>::isValidIdentifier):
(WTF::ObjectIdentifierGenericBase<UUID>::ObjectIdentifierGenericBase):
(WTF::ObjectIdentifierGenericBase<UUID>::hashTableDeletedValue):
(WTF::ObjectIdentifierGeneric::ObjectIdentifierGeneric):
(WTF::add):
(WTF::ObjectIdentifierGenericBaseHash<uint64_t>::hash):
(WTF::ObjectIdentifierGenericBaseHash<uint64_t>::equal):
(WTF::ObjectIdentifierGenericBaseHash<UUID>::hash):
(WTF::ObjectIdentifierGenericBaseHash<UUID>::equal):
(WTF::operator==):
(WTF::operator>):
(WTF::operator>=):
(WTF::operator<):
(WTF::operator<=):
(WTF::ObjectIdentifierGenericBase::isHashTableDeletedValue const): Deleted.
(WTF::ObjectIdentifierGenericBase::isValid const): Deleted.
(WTF::ObjectIdentifierGenericBase::toUInt64 const): Deleted.
(WTF::ObjectIdentifierGenericBase::operator bool const): Deleted.
(WTF::ObjectIdentifierGenericBase::loggingString const): Deleted.
(WTF::ObjectIdentifierGenericBase::isValidIdentifier): Deleted.
(WTF::ObjectIdentifierGenericBase::ObjectIdentifierGenericBase): Deleted.
(WTF::ObjectIdentifierGenericBase::hashTableDeletedValue): Deleted.
(WTF::ObjectIdentifierGenericBaseHash::hash): Deleted.
(WTF::ObjectIdentifierGenericBaseHash::equal): Deleted.
(WTF::ObjectIdentifierGenericBaseStringTypeAdapter::length const): Deleted.
(WTF::ObjectIdentifierGenericBaseStringTypeAdapter::is8Bit const): Deleted.
(WTF::ObjectIdentifierGenericBaseStringTypeAdapter::writeTo const): Deleted.
(WTF::ObjectIdentifierGenericBaseStringTypeAdapter::ObjectIdentifierGenericBaseStringTypeAdapter): Deleted.
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::escapeIDForJSON):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendWithAsyncReply):
(IPC::Connection::sendWithPromisedReply):
(IPC::Connection::sendWithPromisedReplyOnDispatcher):
(IPC::Connection::send):
(IPC::Connection::sendSync):
(IPC::Connection::waitForAndDispatchImmediately):
* Source/WebKit/Platform/IPC/MessageSender.h:
* Source/WebKit/Platform/IPC/MessageSenderInlines.h:
(IPC::MessageSender::send):
(IPC::MessageSender::sendSync):
(IPC::MessageSender::sendWithAsyncReply):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::send):
(IPC::StreamClientConnection::sendWithAsyncReply):
(IPC::StreamClientConnection::sendSync):
(IPC::StreamClientConnection::waitForAndDispatchImmediately):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
(IPC::StreamServerConnection::send):
* Source/WebKit/Shared/AuxiliaryProcess.h:
(WebKit::AuxiliaryProcess::addMessageReceiver):
(WebKit::AuxiliaryProcess::removeMessageReceiver):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::sendWithAsyncReply):
(WebKit::AuxiliaryProcessProxy::send):
(WebKit::AuxiliaryProcessProxy::sendSync):
(WebKit::AuxiliaryProcessProxy::addMessageReceiver):
(WebKit::AuxiliaryProcessProxy::removeMessageReceiver):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::sendToAllProcesses):
* Source/WebKit/UIProcess/RemotePageProxy.h:
(WebKit::RemotePageProxy::sendWithAsyncReply):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::send):
(WebKit::RemoteRenderingBackendProxy::sendSync):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:

Canonical link: https://commits.webkit.org/279402@main
